### PR TITLE
search frontend: add global filter detection for query transformation

### PR DESCRIPTION
**Pull request by rvantonder**
This PR is about the logic that looks at a search query string and then sets URL parameters.

Before, we would set URL parameters for `case` and `patterntype` in the URL depending on query string contents. We would then erase the `field:value` members of the query string so they don't clutter the search bar. We want to preserve this, BUT we don't want to do this unconditionally any more, since subexpressions can contain `case` and `patterntype`, and the previous code didn't detect any of those possibilities. This leads to buggy. behavior in https://github.com/sourcegraph/sourcegraph/issues/13958.

This PR detects whether we can erase filters like `case` or `patterntype`. We can only do so if they are "global", see the first commit's `findGlobalFilter ` comments that checks this.

In the second commit, I use `findGlobalFilter` and simplify our current logic. See inline comments for more. @attfarhan would you throw a careful eye at the changes in this second commit?

### CODE REVIEW

File: client/shared/src/util/url.ts [hunk 1]
```diff
@@ -585,50 +585,32 @@ export function buildSearchURLQuery(
...
-        searchParameters.set('q', fullQuery)
-        searchParameters.set('patternType', patternType)
-    }
-
-    const caseInQuery = parseCaseSensitivityFromQuery(fullQuery)
-    if (caseInQuery) {
-        fullQuery = replaceRange(fullQuery, caseInQuery.range)
-        searchParameters.set('q', fullQuery)
-
-        if (discreteValueAliases.yes.includes(caseInQuery.value)) {
```
Comment (rvantonder) [inline client/shared/src/util/url.ts:610]: As I was simplifying this code, I realized there's a subtle bug here: `caseInQuery.value` could technically be in quotes, so `case:"yes"`. If the value here is `"yes"`, the test `discreteValueAlises.yes.includes('"yes"')` would be false, and we would not evaluate it case sensitively. You can see this bug in action with a query like:

`repo:^github\.com/sourcegraph/sourcegraph$ search case:"yes"`

it will return case insensitive results (a capitalized `search` is in there).

 Basically, this condition wouldn't fire and we would delete the case. Effectively, what this previous code did was assume that the value of `caseInQuery.value` was a `Literal` token.

I've preserved this behavior (did _not_ fix the bug) because that's going to be a distraction right now. I did make it explicit that we are only testing against the `literal` value.

Comment (rvantonder) [reply client/shared/src/util/url.ts:610]: Filed https://github.com/sourcegraph/sourcegraph/issues/16186

File: client/shared/src/util/url.test.ts [hunk 2]
```diff
@@ -342,60 +342,69 @@ describe('withWorkspaceRootInputRevision', () => {
...
 describe('buildSearchURLQuery', () => {
     it('builds the URL query for a regular expression search', () =>
-        expect(buildSearchURLQuery('foo', SearchPatternType.regexp, false, undefined)).toBe('q=foo&patternType=regexp'))
+        expect(buildSearchURLQuery('foo', SearchPatternType.regexp, false, undefined)).toMatchInlineSnapshot(
+            'patternType=regexp&case=no&q=foo'
+        ))
     it('builds the URL query for a literal search', () =>
-        expect(buildSearchURLQuery('foo', SearchPatternType.literal, false, undefined)).toBe(
-            'q=foo&patternType=literal'
+        expect(buildSearchURLQuery('foo', SearchPatternType.literal, false, undefined)).toMatchInlineSnapshot(
```
Comment (rvantonder) [inline client/shared/src/util/url.test.ts:356]: Updated everything to use `toMatchInlineSnapshot` for my sanity.

File: client/shared/src/util/url.test.ts [hunk 3]
```diff
@@ -342,60 +342,69 @@ describe('withWorkspaceRootInputRevision', () => {
...
+            'patternType=literal&case=no&q=foo'
```
Comment (rvantonder) [inline client/shared/src/util/url.test.ts:357]: `case=no` is the default behavior. Let's allow this logic to always set it explicitly.

File: client/shared/src/search/parser/validate.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * For example, `case:yes` is not global in the following queries:
+ *
+ * `(case:yes some subexpression) case:no multiple cases`
+ * `(case:yes not at top level; inside a parentheses of a grouped expression)`
+ *
+ * @param query the query string
+ * @param field the field of the filter to find
+ */
+export const findGlobalFilter = (query: string, field: string): Filter | undefined => {
+    const result = scanSearchQuery(query)
```
Comment (rvantonder) [inline client/shared/src/search/parser/validate.ts:15]: We scan the query string twice with the current code, just like before. We don't need to do this twice in that context, but I will optimize this sort of thing later. Otherwise there's too much movement in one PR.

File: client/shared/src/search/parser/validate.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    let filter: Filter | undefined
+    if (result.type === 'success') {
+        let balanced = 0
```
Comment (attfarhan) [inline client/shared/src/search/parser/validate.ts:18]: It seems like balanced is unused? Or will it serve a purpose down the line?

File: client/shared/src/util/url.ts [hunk 6]
```diff
@@ -585,50 +585,32 @@ export function buildSearchURLQuery(
...
-            // queries breaking when only `repo:` filters are specified.
-            //
-            // TODO: just set case=no when https://github.com/sourcegraph/sourcegraph/issues/7671 is fixed.
-            searchParameters.delete('case')
-        }
+    const globalPatternType = findGlobalFilter(queryParameter, 'patterntype')
+    if (globalPatternType?.value) {
+        const { start, end } = globalPatternType.range
+        queryParameter = replaceRange(queryParameter, { start: Math.max(0, start - 1), end }).trim()
+        patternTypeParameter = query.slice(globalPatternType.value?.range.start)
```
Comment (attfarhan) [inline client/shared/src/util/url.ts:602]: We need the end range passed to `slice`. patternType could be at the start/middle of a query.

File: client/shared/src/util/url.test.ts [hunk 7]
```diff
@@ -342,60 +342,69 @@ describe('withWorkspaceRootInputRevision', () => {
...
-            )
+            expect(
+                buildSearchURLQuery('foo patternType:literal', SearchPatternType.regexp, false, undefined)
+            ).toMatchInlineSnapshot('patternType=literal&case=no&q=foo')
         })
         it('overrides the patternType parameter at the beginning', () => {
             expect(
                 buildSearchURLQuery('patternType:literal foo type:diff', SearchPatternType.regexp, false, undefined)
-            ).toBe('q=foo+type:diff&patternType=literal')
+            ).toMatchInlineSnapshot('patternType=literal+foo+type:diff&case=no&q=foo+type:diff')
```
Comment (attfarhan) [inline client/shared/src/util/url.test.ts:380]: This test gives us the wrong value, and represents the bug from not including the end range when parsing patternType (see my other comment). This would give us a URL `https://sourcegraph.com/search?patternType=literal+foo+type:diff&case=no&q=foo+type:diff`. What we want is `https://sourcegraph.com/search?patternType=literal&case=no&q=foo+type:diff`. This issue appears for a test below too, where patternType is not at the end.

File: client/shared/src/util/url.test.ts [hunk 8]
```diff
@@ -342,60 +342,69 @@ describe('withWorkspaceRootInputRevision', () => {
...
             expect(
                 buildSearchURLQuery('type:diff foo patternType:literal', SearchPatternType.regexp, false, undefined)
-            ).toBe('q=type:diff+foo&patternType=literal')
+            ).toMatchInlineSnapshot('patternType=literal&case=no&q=type:diff+foo')
         })
         it('overrides the patternType parameter in the middle', () => {
             expect(
                 buildSearchURLQuery('type:diff patternType:literal foo', SearchPatternType.regexp, false, undefined)
-            ).toBe('q=type:diff+foo&patternType=literal')
+            ).toMatchInlineSnapshot('patternType=literal+foo&case=no&q=type:diff+foo')
```
Comment (attfarhan) [inline client/shared/src/util/url.test.ts:390]: Same issue as other test

File: client/shared/src/util/url.ts [hunk 9]
```diff
@@ -585,50 +585,32 @@ export function buildSearchURLQuery(
...
+    searchParameters.set('patternType', patternTypeParameter)
+
+    const globalCase = findGlobalFilter(queryParameter, 'case')
+    if (globalCase?.value && globalCase.value.type === 'literal') {
+        caseParameter = globalCase.value.value
+        queryParameter = replaceRange(queryParameter, globalCase.range)
+    }
+    searchParameters.set('case', caseParameter)
+
+    searchParameters.set('q', queryParameter)
```
Comment (attfarhan) [inline client/shared/src/util/url.ts:613]: Unsure if this is deliberate, but I believe this orders the URL to display the parameters in the order: patternType, case, query. Right now we show the query first, and I think it would be good to maintain this. Wdyt?

**Review by attfarhan — CHANGES_REQUESTED** [review]

File: client/shared/src/search/parser/validate.ts (hunk 5 above)
Comment (rvantonder) [reply client/shared/src/search/parser/validate.ts:18]: Oh good catch! I don't think we need this right now.

File: client/shared/src/util/url.ts (hunk 9 above)
Comment (rvantonder) [reply client/shared/src/util/url.ts:613]: Changing the order change isn't deliberate. It's just that the previous code runs this call 4 times inside a nested conditionals and it's avoidable. Since order isn't significant, I didn't pay attention to that. I'll reorder the `set` calls so that the previous order is preserved.

File: client/shared/src/util/url.ts (hunk 6 above)
Comment (rvantonder) [reply client/shared/src/util/url.ts:602]: facepalm

thankssss

**Comment by rvantonder** [comment]
@attfarhan I changed the code a bit so that we preserve the original URL param order and whether we put `case` in the URL params, such that no url tests need to be updated any more. I've undone the previous test changes. Should be good now.

**Review by attfarhan — APPROVED** [review]
